test/basic_test.go

```diff
@@ -2683,6 +2683,145 @@ func TestViewChangeAfterTryingToFork(t *testing.T) {
 	}
 }
 
+func TestFetchStateWhenSyncReturnsPrevView(t *testing.T) {
+	// Scenario: A node (node 2) becomes the leader but disconnects inducing a second view change.
+	// When it reconnects, with no transactions available, it syncs after receiving multiple heartbeat responses.
+	// However, the Synchronizer returns with an old view, and so for the node to catch up it needs to fetch the state.
+
+	t.Parallel()
+	network := NewNetwork()
+	defer network.Shutdown()
+
+	testDir, err := os.MkdirTemp("", t.Name())
+	assert.NoErrorf(t, err, "generate temporary test dir")
+	defer os.RemoveAll(testDir)
+
+	numberOfNodes := 4
+	nodes := make([]*App, 0)
+	for i := 1; i <= numberOfNodes; i++ {
+		n := newNode(uint64(i), network, t.Name(), testDir, false, 0)
+		nodes = append(nodes, n)
+	}
+
+	start := time.Now()
+	for _, n := range nodes {
+		n.heartbeatTime = make(chan time.Time, 1)
+		n.heartbeatTime <- start
+		n.Setup()
+	}
+
+	// wait for the new leader to finish the view change before submitting
+	done := make(chan struct{})
+	viewChangeWG := sync.WaitGroup{}
+	viewChangeWG.Add(numberOfNodes - 1)
+
+	syncWG := sync.WaitGroup{}
+	syncWG.Add(1)
+
+	ignoreWG := sync.WaitGroup{}
+	ignoreWG.Add(1)
+	ignoreOnce := sync.Once{}
+
+	stateWG := sync.WaitGroup{}
+	stateWG.Add(1)
+
+	for _, n := range nodes {
+		id := n.ID
+		baseLogger := n.Consensus.Logger.(*zap.SugaredLogger).Desugar()
+		n.Consensus.Logger = baseLogger.WithOptions(zap.Hooks(func(entry zapcore.Entry) error {
+			if strings.Contains(entry.Message, "ViewChanged") {
+				viewChangeWG.Done()
+			}
+
+			if strings.Contains(entry.Message, "Received HeartBeatResponse triggered a call to HeartBeatEventHandler Sync") {
+				syncWG.Done()
+			}
+
+			if strings.Contains(entry.Message, "Monitor already called Sync, ignoring HeartBeatResponse") {
+				if id == 2 {
+					ignoreOnce.Do(func() {
+						ignoreWG.Done()
+					})
+				}
+			}
+
+			if strings.Contains(entry.Message, "Collected state with view 2 and sequence 2") {
```

Comment:
I'm not sure I agree to the fact that this should be the trigger to end the test.
I think we should instead proceed to send a new transaction afterwards and make sure that this node participated in consensus.

Reply:
I can just move the wait part before the submit of the transaction

Reply:
yeah makes more sense than what i suggested!